examples/multicast.go

```diff
@@ -0,0 +1,185 @@
+	return zap.S()
+}
+
+func saveData(data interface{}, filename string) error {
+	f, err := os.Create(*storagePath + filename)
+	if err != nil {
+		log.Error("Fail to create file", "filename", filename, "error", err)
+		return err
+	}
+
+	err = json.NewEncoder(f).Encode(data)
+	if err != nil {
+		log.Error("Fail to write data to file", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// listen to multicast orderbook events
+func listenToOrderbookEvent(ctx context.Context, m *multicast.Client) {
+
+	orderbookChannels := []string{
+		"book.BTC-PERPETUAL",
+		"book.BTC-1AUG22-29000-P",
+	}
+	data := make([]models.OrderBookRawNotification, 0)
+	listener := func(e *models.OrderBookRawNotification) {
+		data = append(data, *e)
+	}
+	for _, channel := range orderbookChannels {
+		m.On(channel, listener)
+
+	}
+
+	<-ctx.Done()
+
+	for _, channel := range orderbookChannels {
+		m.Off(channel, listener)
+	}
+	saveData(data, "orderbook.json")
+}
+
+// listen to multicast trades events
+func listenToTradesEvent(ctx context.Context, m *multicast.Client) {
+	tradesChannels := []string{
+		"trade.option.BTC",
+		"trade.future.BTC",
+	}
+	data := make([]models.TradesNotification, 0)
+	listener := func(e *models.TradesNotification) {
+		data = append(data, *e)
+	}
+	for _, channel := range tradesChannels {
+		m.On(channel, listener)
+
+	}
+
+	<-ctx.Done()
+
+	for _, channel := range tradesChannels {
+		m.Off(channel, listener)
+	}
+
+	saveData(data, "trades.json")
+}
+
+// listen to multicast ticker events
+func listenToTickerEvent(ctx context.Context, m *multicast.Client) {
+	tickerChannels := []string{
+		"ticker.BTC-PERPETUAL",
+		"ticker.BTC-1AUG22-29000-P",
+	}
+	data := make([]models.TickerNotification, 0)
+	listener := func(e *models.TickerNotification) {
+		data = append(data, *e)
+	}
+	for _, channel := range tickerChannels {
+		m.On(channel, listener)
+
+	}
+
+	<-ctx.Done()
+
+	for _, channel := range tickerChannels {
+		m.Off(channel, listener)
+	}
+	saveData(data, "ticker.json")
+}
+
+func main() {
+	wsConfig := &ws.Configuration{
+		Addr:          *wsEndpoint,
+		ApiKey:        *apiKey,
+		SecretKey:     *secretKey,
+		AutoReconnect: true,
+		DebugMode:     true,
+	}
+	log = setupLogger(*debug)
+
+	wsClient := ws.New(log, wsConfig)
+	udpAddrs := strings.Split(*addrs, ",")
+	multicastClient, err := multicast.NewClient(*ifname, udpAddrs, wsClient, []string{"BTC"})
+	if err != nil {
+		log.Errorw("failed to initiate multicast client", "ifname", ifname, "addrs", addrs)
+		panic(err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		time.Sleep(*gatherDataDuration)
+		cancel()
+	}()
+
+	err = multicastClient.Start(ctx)
+	if err != nil {
+		log.Errorw("failed to start multicast client", "ifname", ifname, "addrs", addrs)
+		panic(err)
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(3) // for orderbook, trades, ticker notifications
+
+	go func() {
+		defer wg.Done()
+		listenToOrderbookEvent(ctx, multicastClient)
```

Comment:
Use context with timeout for this function, don't use the time context pass to Start.